cmd/iam-k8s-store.go

```diff
@@ -0,0 +1,554 @@
+				Name:            iamK8s.configMapName,
+				ResourceVersion: configMap.ResourceVersion,
+				Annotations:     annotations,
+			},
+		}
+		_, err = iamK8s.configMapsClient.Update(ctx, configMapUpdated, metav1.UpdateOptions{})
+		if err != nil {
+			if errorsv1.IsConflict(err) {
+				attempts += 1
+				continue
+			} else if isRetryable(err) {
+				attempts += 1
+				backoffDefault(attempts)
+				continue
+			} else {
+				return err
+			}
+		} else {
+			return nil
+		}
+	}
+	return nil
+}
+
+func (iamK8s *IAMK8sStore) loadIAMConfig(ctx context.Context, item interface{}, objPath string) error {
+	configMap, err := iamK8s.configMapsClient.Get(ctx, iamK8s.configMapName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	data := configMap.Annotations[objPathToK8sValid(objPath)]
+	if data == "" {
+		return errors.New("No entry for key: " + objPath)
+	}
+	return json.Unmarshal([]byte(data), item)
+}
+
+type iamConfigItem struct {
+	objPath string
+	data    string
+}
+
+type iamConfigListResult struct {
+	iamConfigItems  []iamConfigItem
+	resourceVersion string
+	err             error
+}
+
+func (iamK8s *IAMK8sStore) listIAMConfigs(ctx context.Context, pathPrefix string, includeTtlItems bool) iamConfigListResult {
+	configItems := []iamConfigItem{}
+	configMap, err := iamK8s.configMapsClient.Get(ctx, iamK8s.configMapName, metav1.GetOptions{})
+	if err != nil {
+		return iamConfigListResult{err: err}
+	}
+
+	for objPath, data := range configMap.Annotations {
+		objPath = objPathFromK8sValid(objPath)
+		trimmedObjPath := strings.TrimPrefix(objPath, pathPrefix)
+		trimmedObjPath = strings.TrimSuffix(trimmedObjPath, SlashSeparator)
+		ttlOk := (includeTtlItems && strings.HasPrefix(trimTtlPrefix(objPath), pathPrefix))
+		if strings.HasPrefix(objPath, pathPrefix) || ttlOk {
+			configItems = append(configItems, iamConfigItem{objPath: trimmedObjPath, data: data})
+		}
+	}
+
+	return iamConfigListResult{configItems, configMap.ResourceVersion, nil}
+}
+
+func (iamK8s *IAMK8sStore) deleteIAMConfig(ctx context.Context, path string) error {
+	configMap, err := iamK8s.configMapsClient.Get(ctx, iamK8s.configMapName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	annotations := configMap.Annotations
+	if _, ok := annotations[path]; ok {
+		return errConfigNotFound
+	} else {
+		delete(annotations, path)
+		configMapUpdated := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:       iamK8s.namespace,
+				Name:            iamK8s.configMapName,
+				ResourceVersion: configMap.ResourceVersion,
+				Annotations:     annotations,
+			},
+		}
+		_, err = iamK8s.configMapsClient.Update(ctx, configMapUpdated, metav1.UpdateOptions{})
+		return err
+	}
+}
+
+func (iamK8s *IAMK8sStore) loadPolicyDoc(ctx context.Context, policy string, m map[string]iampolicy.Policy) error {
+	var p iampolicy.Policy
+	err := iamK8s.loadIAMConfig(ctx, &p, getPolicyDocPath(policy))
+	if err != nil {
+		if err == errConfigNotFound {
```

Comment:
Would perhaps wrapping the error be better? Also I would make error instances (errNoSuchPolicy) public such that people can match against them, and programmatically deal with different scenarios, in general.

But that only applies if users will ever see error results from this method, or any private methods in this package/struct.